### PR TITLE
Add filtration feature in the bulk API

### PR DESCRIPTION
**Pull request by khansaad**
## Description

This PR adds the filtration feature in the bulk API using which user can hit bulk API with filtered resources like namespaces, containers and workloads instead of always fetching everything.

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Docs update
- [ ] Breaking change (What changes might users need to make in their application due to this PR?)
- [ ] Requires DB changes

## How has this been tested?

By manually hitting bulk API with multiple combinations of include and exclude filters

Please describe the tests that were run to verify your changes and steps to reproduce. Please specify any test configuration required. 

- [ ] New Test X
- [x] Functional testsuite

**Test Configuration**
* Kubernetes clusters tested on: Openshift

## Checklist :dart:

- [x] Followed coding guidelines
- [x] Comments added
- [x] Dependent changes merged
- [ ] Documentation updated
- [ ] Tests added or updated

## Additional information

Include any additional information such as links, test results, screenshots here

### CODE REVIEW

**Comment by dinogun** [comment]
@khansaad This needs a Documentation update to mention the new filters being added. Also please add a few example filters in the kruize-demos repo (Eg Include only openshift-* namespaces or exclude them etc) and test them and post the results here, thanks

**Comment by khansaad** [comment]
[bulk-filter-demo-results.zip](https://github.com/user-attachments/files/18123615/bulk-filter-demo-results.zip)


> @khansaad This needs a Documentation update to mention the new filters being added. Also please add a few example filters in the kruize-demos repo (Eg Include only openshift-* namespaces or exclude them etc) and test them and post the results here, thanks

**Comment by dinogun** [comment]
@khansaad Please fix the conflicts

**Comment by khansaad** [comment]
> 

Done

File: design/BulkAPI.md [hunk 1]
```diff
@@ -28,18 +28,23 @@ progress of the job.
...
+        "cluster_id": "ABG"
+      }
     },
     "include": {
-      "namespace": [],
-      "workload": [],
-      "containers": [],
+      "namespace": ["cadvisor", "openshift-tuning", "openshift-monitoring", "thanos-bench"],
+      "workload": ["osd-rebalance-infra-nodes-28887030", "blackbox-exporter", "thanos-query"],
+      "containers": ["tfb-0", "alertmanager"],
```
Comment (dinogun) [inline design/BulkAPI.md:43]: This is a bad example, the same things are present in both the `include` and `exclude` criteria. Also are not supporting wildcard chars here?

What we need to support is something like this
```
"exclude": {
   "namespace": ["openshift-*"]
},
"include": {
   "namespace": ["openshift-tuning"]
}
```

This should exclude all "openshift-" namespaces except "openshift-tuning"

**Comment by dinogun** [comment]
With this bulk_input.json
```
{
  "filter": {
    "exclude": {
      "namespace": ["openshift-.*"]
    }
  },
  "datasource": "prometheus-1"
}
```

I see that bulk API thinks there are 0 experiments
```
$ ./bulk_service_demo.sh -c openshift -i quay.io/dinogun210/autotune_operator:0.2-pr1391

#######################################
# Kruize Demo Setup on openshift 
#######################################

🔄 Pulling required repositories... ✅ Done!
🔄 Installing kruize! Please wait..............✅ Installation of kruize complete!
🔄 Installing metric profile...✅ Installation of metric profile complete!
🔄 Invoking bulk service...✅ Invoked job_id 32b02bb5-cbed-447d-a767-704ff073da5a
🔔 Bulk job experiment details are currently being updated in ./kruize-bulk-demo.log.
🔄 Processing 0 experiments. Please wait....✅ Complete!
🔄 Fetching the experiments...✅ Complete!
🔄 List the recommendations for 1 experiments...✅ Complete!
📌 List of all experiments available in experiment_list.txt
📌 Recommendations for a single container in cluster can be found in recommendations_data.json.
📌 Job status is available in job_status.json

```

**Comment by khansaad** [comment]
> With this bulk_input.json
> 
> ```
> {
>   "filter": {
>     "exclude": {
>       "namespace": ["openshift-.*"]
>     }
>   },
>   "datasource": "prometheus-1"
> }
> ```
> 
> I see that bulk API thinks there are 0 experiments
> 
> ```
> $ ./bulk_service_demo.sh -c openshift -i quay.io/dinogun210/autotune_operator:0.2-pr1391
> 
> #######################################
> # Kruize Demo Setup on openshift 
> #######################################
> 
> 🔄 Pulling required repositories... ✅ Done!
> 🔄 Installing kruize! Please wait..............✅ Installation of kruize complete!
> 🔄 Installing metric profile...✅ Installation of metric profile complete!
> 🔄 Invoking bulk service...✅ Invoked job_id 32b02bb5-cbed-447d-a767-704ff073da5a
> 🔔 Bulk job experiment details are currently being updated in ./kruize-bulk-demo.log.
> 🔄 Processing 0 experiments. Please wait....✅ Complete!
> 🔄 Fetching the experiments...✅ Complete!
> 🔄 List the recommendations for 1 experiments...✅ Complete!
> 📌 List of all experiments available in experiment_list.txt
> 📌 Recommendations for a single container in cluster can be found in recommendations_data.json.
> 📌 Job status is available in job_status.json
> ```


Fixed the issue with experiments creation now but the demo is failing at `listRecommendations` step which is expected because of the recent DB changes. As @msvinaykumar mentioned, this should work after #1420 and #1421 merge

**Comment by khansaad** [comment]
`saakhan:bulk_demo$ ./bulk_service_demo.sh -c openshift -i quay.io/khansaad/autotune_operator:bulk-filter


#######################################
# Kruize Demo Setup on openshift
#######################################

🔄 Pulling required repositories... ✅ Done!
🔄 Installing kruize! Please wait....................✅ Installation of kruize complete!
Running bulk_demo.py...
bulk_demo.py -c openshift
Error from server (AlreadyExists): routes.route.openshift.io "kruize" already exists
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
IP =  kruize-openshift-tuning.apps.kruize2.lab.psi.pnq2.redhat.com

KRUIZE URL =  http://kruize-openshift-tuning.apps.kruize2.lab.psi.pnq2.redhat.com

#######################################
Creating metric profile
#######################################

URL =  http://kruize-openshift-tuning.apps.kruize2.lab.psi.pnq2.redhat.com/createMetricProfile
Response status code =  201
{"message":"Metric Profile : resource-optimization-local-monitoring created successfully. View Metric Profiles at /listMetricProfiles","httpcode":201,"documentationLink":"","status":"SUCCESS"}

#######################################
Invoking bulk service
#######################################

URL =  http://kruize-openshift-tuning.apps.kruize2.lab.psi.pnq2.redhat.com/bulk
Response -  {'job_id': '730e3f0b-2ed6-4dfe-a962-1da96273fc31'}

#######################################
Querying job status in a loop
#######################################

Experiments: processed / Total -  0 / 0
Experiments: processed / Total -  0 / 8
Experiments: processed / Total -  3 / 8
Experiments: processed / Total -  6 / 8
Experiments: processed / Total -  8 / 8

#######################################
Bulk Job Completed! Fetching the processed experiments and listing recommendations
#######################################`

**Comment by chandrams** [comment]
@msvinaykumar - Can you please review this PR

**Comment by chandrams** [comment]
@khansaad - Tested with the below input json with [quay.io/khansaad/autotune_operator:bulk-filter-with-cft](http://quay.io/khansaad/autotune_operator:bulk-filter-with-cft), but it shows 0 processed experiments. 

```
{
  "filter": {
    "exclude": {
      "namespace": ["openshift-.*"],
      "workload": [],
      "containers": [],
      "labels": {}
    },
    "include": {
      "namespace": ["openshift-tuning"],
      "workload": [],
      "containers": [],
      "labels": {}
    }
  },

  "time_range": {
        "start": null,
        "end": null
  }
```

Job status
```
{
    "status": "COMPLETED",
    "total_experiments": 0,
    "processed_experiments": 0,
    "notifications": {
        "200": {
            "type": "INFO",
            "message": "Nothing to do.",
            "code": 400
        }
    },
    "experiments": {},
    "webhook": null,
    "job_id": "599b70bb-a1a6-4814-96a0-7bfd5fb80e81",
    "job_start_time": "2024-12-17T11:48:52.071Z",
    "job_end_time": "2024-12-17T11:48:52.115Z"
}
```

**Review by dinogun — APPROVED** [review]
LGTM

**Pull request merged**